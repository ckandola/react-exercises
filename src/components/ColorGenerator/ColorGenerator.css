/*
=============== 
Variables
===============
*/

:root {
    --clr-primary-5: hsl(205, 78%, 60%);
    /* darkest grey - used for headings */
    --clr-grey-1: hsl(209, 61%, 16%);
    --clr-white: #fff;
    --clr-red-dark: hsl(360, 67%, 44%);
    --radius: 0.25rem;
    --max-width: 1170px;
  }

  /* section */
  .cgen-section {
    width: 90vw;
    margin: 0 auto;
    max-width: var(--max-width);
  }
  
  @media screen and (min-width: 992px) {
    .cgen-section {
      width: 95vw;
    }
  }
  
  /*
  =============== 
  Color Generator
  ===============
  */
  
  .cgen-container {
    text-align: center;
    display: flex;
    align-items: center;
    height: 100px;
    padding-left: 2rem;
  }
  .cgen-container h3 {
    margin-bottom: 0;
    margin-right: 2rem;
  }
  
  .cgen-input {
    border-color: transparent;
    width:15vw;
    padding: 0.5rem 1rem;
    font-size: 1.2rem;
    border-top-left-radius: var(--radius);
    border-bottom-left-radius: var(--radius);
    margin-left: 1rem;
  }
  .cgen-input-num {
    width: 4em;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }

  .cgen-color-btn {
    width: 2em;
    height: 2em;
    margin: 0.25em;
    display: inline-grid;
    border-radius: 0.5em;
    cursor: pointer;
  }

  .cgen-color-picker {
    display: block;
    text-align: center;
    background-color: #A0A0A0;
    border-radius: 0.7em;
    margin: auto;
    max-width: 35%;
    padding:0.25em;
    margin-bottom: 1em;
  }

  .cgen-btn {
    background: var(--clr-primary-5);
    padding: 0.5rem 1rem;
    font-size: 1.2rem;
    border-color: transparent;
    border-top-right-radius: var(--radius);
    border-bottom-right-radius: var(--radius);
    text-transform: capitalize;
    color: var(--clr-white);
    cursor: pointer;
    margin-left: 1rem;
  }
  @media screen and (max-width: 576px) {
    .cgen-container h3 {
      font-size: 1rem;
    }
    .cgen-input,
    .cgen-btn {
      font-size: 0.85rem;
    }
  }
  .cgen-input-error {
    border: 2px solid var(--clr-red-dark);
  }
  .cgen-colors {
    min-height: calc(100vh - 100px);
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(223.33px, 1fr));
    grid-template-rows: repeat(auto-fit, minmax(96px, 1fr));
  }
  .cgen-color {
    padding: 1rem 2rem;
    cursor: pointer;
    font-size: 1rem;
    text-transform: none;
  }
  .cgen-percent-value {
    margin-bottom: 0;
    color: var(--clr-grey-1);
  }
  .cgen-color-value {
    color: var(--clr-grey-1);
    margin-bottom: 0;
  }
  .cgen-color-light .color-value {
    color: var(--clr-white);
  }
  .cgen-color-light .percent-value {
    color: var(--clr-white);
  }
  
  .cgen-alert {
    text-transform: uppercase;
    font-size: 0.85rem;
    margin-top: 0.5rem;
  }
