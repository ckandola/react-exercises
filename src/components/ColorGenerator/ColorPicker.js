import React from 'react';
import './ColorGenerator.css';

const colorList = [
    '#000000',  // black
    '#C0C0C0',  // gray
    '#804040',  // brown
    '#B5744D',  // tan
    '#FEDBB8',  // beige
    '#5600AC',  // violet
    '#AC77F9',  // periwinkle
    '#0000ff',  // blue
    '#0080ff',  // sky blue
    '#00f0ff',  // cyan
    '#008000',  // dark green
    '#13B517',  // green
    '#00ff00',  // light green
    '#eedc02',  // gold
    '#ffff00',  // yellow
    '#ff8000',  // orange
    '#ffA851',  // light orange
    '#800000',  // burgundy
    '#ff0000',  // red
    '#ff80C0',  // pink
    '#ff00ff',  // magenta
    '#ffffff'   // white
];


const ColorPicker = ({colors = colorList, onClick}) => {
    const handleClick = chosenColor => {
        onClick(chosenColor);
    }

    return (
        <section className="cgen-color-picker">
            <div>
                {colors.map((color, index) => {
                    return <div key={index} className="cgen-color-btn" style={{backgroundColor: color}} onClick={() => handleClick(color)} />;
                })}
            </div>
        </section>
    );
};

export default ColorPicker;
