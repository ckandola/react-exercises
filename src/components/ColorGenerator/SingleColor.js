import React, {useState, useEffect} from 'react';
import './ColorGenerator.css';
import PropTypes from 'prop-types';

const SingleColor = ({rgb, weight, index, hexColor}) => {
    const [alert, setAlert] = useState(false);
    const bgcolor = rgb.join(',');

    useEffect(() => {
        const hideText = setTimeout(() => {
            setAlert(false);
        }, 3000);
        return () => clearTimeout(hideText);
    }, [alert]);

    return (
        <div 
            className={`cgen-color ${index > 10 && 'cgen-color-light'}`} 
            style={{backgroundColor: `rgb(${bgcolor})`}}
            onClick={() => {
                setAlert(true);
                navigator.clipboard.writeText(`#${hexColor}`);
            }}
        >
            <p className="cgen-percent-value">{weight}%</p>
            <p className="cgen-color-value">#{hexColor}</p>
            {alert && <p className="cgen-alert">
                Copied to clipboard</p>}
        </div>
    );
};

export default SingleColor;

SingleColor.propTypes = {
    rgb: PropTypes.arrayOf(PropTypes.number),
    weight: PropTypes.number,
    index: PropTypes.number,
    hexColor: PropTypes.string
};
