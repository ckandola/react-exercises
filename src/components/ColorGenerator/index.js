// from https://youtu.be/a_7Z7C_JCyo?t=11478

import React, {useState} from 'react';
import './ColorGenerator.css';
import Values from 'values.js';
import SingleColor from './SingleColor';

const ColorGenerator = () => {
    const [color, setColor] = useState('#8b0000');
    const [error, setError] = useState(false);
    const [list, setList] = useState([]);
    const [numResults, setNumResults] = useState(10);

    const handleSubmit = e => {
        e.preventDefault();
        try {
            // generates shades and tints divided by 10, 100 / 10 = 10. Larger number ==> less shades & tints
            let colors = new Values(color).all(numResults);
            setList(colors);
        } catch (error) {
            setError(true);
            console.log(error);
        }
    };

    const handleNumChange = e => {
        if (!e.target.value) {
            return;
        }
        setNumResults(e.target.value <= 0 ? 1 : Math.floor(100 / Math.floor(e.target.value)));
    };

    return (
        <>
            <div className="cgen-container">
                <h3>Color Generator</h3>
                <form onSubmit={handleSubmit}>
                    <input 
                        type="text" 
                        value={color}
                        onChange={e => setColor(e.target.value)}
                        className={`${error ? 'error' : 'null'}`}/>
                    <input
                        type="number"
                        className="cgen-input-num"
                        placeholder={Math.floor(100 / numResults)}
                        step={1}
                        onChange={e => handleNumChange(e)}
                    />
                    <button className="cgen-btn" type="submit">Submit</button>
                </form>
            </div>
            <section className="cgen-colors">
                {list.map((listItem, index) => {
                    return (
                        <SingleColor key={index} index={index} className="cgen-color" {...listItem} hexColor={listItem.hex} />
                    );
                })}
            </section>
        </>
    );
};

export default ColorGenerator;
