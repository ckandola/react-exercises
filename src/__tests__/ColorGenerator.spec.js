import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ColorGenerator from '../components/ColorGenerator';

describe('ColorGenerator tests', () => {
    
    Object.assign(navigator, {
        clipboard: {
            writeText: jest.fn().mockImplementation(() => Promise.resolve()),
        },
    });

    const wrapper = mount(<ColorGenerator />);

    test('matches the snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('changes contents of input to `#0000ff` when blue color button is clicked', () => {
        wrapper.find({style: {backgroundColor: '#0000ff'}}).simulate('click');
        expect(wrapper.find('.cgen-input').at(0).getElement().props.value).toBe('#0000ff');
    });

    test('when second input is left at `10`, clicking Submit provides 21 results', () => {
        expect(wrapper.find('.cgen-input').at(1).getElement().props.placeholder).toBe(10);
        wrapper.find('form').simulate('submit', { preventDefault () {} });
        expect(wrapper.find('SingleColor')).toHaveLength(21);
    });

    test('setting the numeric input to a negative number results in 201 results', () => {
        wrapper.find('.cgen-input').at(1).simulate('change', {target: {value: -1}});
        wrapper.find('form').simulate('submit', { preventDefault () {} });
        expect(wrapper.find('SingleColor')).toHaveLength(201);
        wrapper.find('.cgen-input').at(1).simulate('change', {target: {value: 10}});
        wrapper.find('form').simulate('submit', { preventDefault () {} });
    });
    
    test('clicking a SingleColor triggers a message, `Copied to clipboard`', () => {
        wrapper.find('SingleColor').at(5).simulate('click');
        expect(wrapper.find('.cgen-alert').getElement().props.children).toBe("Copied to clipboard");
    });
});
