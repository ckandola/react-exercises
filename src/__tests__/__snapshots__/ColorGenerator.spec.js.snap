// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorGenerator tests matches the snapshot 1`] = `
<ColorGenerator>
  <div
    className="cgen-container"
  >
    <h3>
      Color Generator
    </h3>
    <form
      onSubmit={[Function]}
    >
      <input
        className="cgen-input null"
        onChange={[Function]}
        type="text"
        value="#8b0000"
      />
      <input
        className="cgen-input cgen-input-num"
        onChange={[Function]}
        placeholder={10}
        step={1}
        type="number"
      />
      <button
        className="cgen-btn"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <ColorPicker
    onClick={[Function]}
  >
    <section>
      <div
        className="cgen-color-picker"
      >
        <div
          className="cgen-color-btn"
          key="0"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#000000",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="1"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#C0C0C0",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="2"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#804040",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="3"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#B5744D",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="4"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#FEDBB8",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="5"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#5600AC",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="6"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#AC77F9",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="7"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#051a8d",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="8"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#0000ff",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="9"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#0080ff",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="10"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#00f0ff",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="11"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#23af89",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="12"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#008000",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="13"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#13B517",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="14"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#00ff00",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="15"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#eedc02",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="16"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ffff00",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="17"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ff8000",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="18"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ffA851",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="19"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#800000",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="20"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ff0000",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="21"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ff80C0",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="22"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ff00ff",
            }
          }
        />
        <div
          className="cgen-color-btn"
          key="23"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "#ffffff",
            }
          }
        />
      </div>
    </section>
  </ColorPicker>
  <section
    className="cgen-colors"
  />
</ColorGenerator>
`;
